### PR TITLE
preassembled_seed_fragmentation

### PR DESCRIPTION
**Pull request by pb-cdunn**
* https://jira.pacificbiosciences.com/browse/SAT-105

### CODE REVIEW

File: falcon_kit/stats_preassembly.py [hunk 1]
```diff
@@ -115,7 +116,25 @@ def read_lens_from_db(db_fn):
     """
     return list(sorted(get_db_readlengths(db_fn)))
 
-def stats_dict(stats_raw_reads, stats_seed_reads, stats_corrected_reads, genome_length, length_cutoff):
+def metric_fragmentation(preads_dir):
+    # https://jira.pacificbiosciences.com/browse/SAT-105
+    bash = r"""
+sed -nr 's;>prolog/([0-9]*)[0-9]/.*;\1;p' %s/*.fasta | sort | uniq -c | awk '{print $1}' | sort | uniq -c
```
Comment (pbjd) [inline falcon_kit/stats_preassembly.py]: That sed command I gave you is pretty inefficient. Why not just do this in pure python with a hashtable?

Comment (pb-cdunn) [inline falcon_kit/stats_preassembly.py]: That's fine with me. You're worried about the runtime of the first sort? How slow was it? Did you measure it in isolation, or is it possible that sed on the gigantic files was actually the slow part?

sed is much faster than python at parsing. If speed is the issue, then perl is a better choice than python. Any objection?

Comment (pbjd) [inline falcon_kit/stats_preassembly.py]: The sed command itself is not what I'm thinking about, it's all the merging and sorting after the pipes that makes me cringe.  It's not slow on bacteria, but will be really slow on larger genomes.

Comment (pb-cdunn) [inline falcon_kit/stats_preassembly.py]: Ok. Maybe you're right. Won't take long to change...

File: falcon_kit/stats_preassembly.py [hunk 2]
```diff
@@ -115,7 +117,20 @@ def read_lens_from_db(db_fn):
     """
     return list(sorted(get_db_readlengths(db_fn)))
 
-def stats_dict(stats_raw_reads, stats_seed_reads, stats_corrected_reads, genome_length, length_cutoff):
+def metric_fragmentation(preads_dir):
+    # https://jira.pacificbiosciences.com/browse/SAT-105
+    #sed -nr 's;>prolog/([0-9]*)[0-9]/.*;\1;p' %s/*.fasta | sort | uniq -c | awk '{print $1}' | sort | uniq -c
+    fastas = ' '.join(glob.glob(preads_dir + '/*.fasta'))
+    call = """perl -e 'while (<>) { if ( m{>[^/]+/(\d+)\d/} ) { $id{$1}++; } }; while (my ($k, $v) = each %%id) { $counts{$v}++; }; while (my ($k, $v) = each %%counts) { print "$v $k\n"; };' %s""" %(fastas)
```
Comment (pbjd) [inline falcon_kit/stats_preassembly.py]: Perl is so awesome at this sort of thing.  Sometimes I actually miss it.

**Comment by mhsieh** [comment]
You need this perl : private/thirdparty/perl/perl_5.24.0/binwrap/perl

File: falcon_kit/stats_preassembly.py (hunk 2 above)
Comment (pbjd) [inline falcon_kit/stats_preassembly.py]: One thing to double check: While perl is in our 3.x.x distro, I'm not sure it's available at runtime.  We can't necessarily assume it's installed at the customer site.

**Comment by pb-cdunn** [comment]
@mhsieh, we *could* add `perl` into the mobs dependencies. But perl is actually everywhere that we might run, and we do not depend on the version.

If mobs used a simpler "module" system, I would be less opposed. But as it is, I'd rather rely on either Linux guarantees or a static binary (as from Go or Nim). I'm really tired of the dependency hell.

**Pull request merged**